File: apps/android/app/src/main/java/com/kris99/baozi/android/ui/common/AgentRuntimePresentation.kt
package com.kris99.baozi.android.ui.common

import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import com.kris99.baozi.android.ui.BaoziTheme
import uniffi.codex_mobile_client.AppAgentMetadata

/**
 * Bridge alias: Rust exposes agent identity as an opaque `String` (the
 * lowercase id alleycat advertises). The legacy `AgentRuntimeKind`
 * name is preserved as a type alias so call sites compile; ALL agent
 * metadata — label, icon, BETA badge, sort order, capability flags —
 * comes from `AgentMetadataStore` keyed by id. There is no hardcoded
 * catalog of agent names in baozi, so adding a new agent only
 * requires an entry in the alleycat manifest.
 */
typealias AgentRuntimeKind = String

/**
 * Lookup hook into the Rust-owned `AgentMetadataStore`. Wired up at
 * app launch in `BaoziApplication`. Returns `null` before the first
 * probe response has populated the cache.
 */
object AgentRuntimeMetadataProvider {
    var lookup: ((String) -> AppAgentMetadata?)? = null
    var all: (() -> List<AppAgentMetadata>)? = null
}

val AgentRuntimeKind.metadata: AppAgentMetadata?
    get() = AgentRuntimeMetadataProvider.lookup?.invoke(this)

/** Short label used in lists. Falls back to titlecased id on cold start. */
val AgentRuntimeKind.runtimeLabel: String
    get() = metadata?.displayName?.takeIf { it.isNotEmpty() } ?: titlecased()

/** Header / title rendering. Prefers metadata `presentation.title`. */
val AgentRuntimeKind.titleDisplayLabel: String
    get() = metadata?.presentation?.title?.takeIf { it.isNotEmpty() } ?: runtimeLabel

/**
 * Ascending sort key from `presentation.sort_order`; unknown agents
 * drop to the end.
 */
val AgentRuntimeKind.runtimeSortIndex: Int
    get() = metadata?.presentation?.sortOrder?.toInt() ?: Int.MAX_VALUE

/**
 * BETA badge driven by `presentation.is_beta`. Codex is always stable,
 * including cold-start SSH/alleycat paths before metadata is cached; other
 * unknown agents stay beta by default until metadata says otherwise.
 */
val AgentRuntimeKind.isBeta: Boolean
    get() = if (isStableAgentIdentity(this, "")) {
        false
    } else {
        metadata?.presentation?.isBeta ?: true
    }

/** Whether this runtime accepts client-side thread permission overrides. */
val AgentRuntimeKind.supportsThreadPermissionOverrides: Boolean
    get() = metadata?.capabilities?.supportsThreadPermissionOverrides ?: true

/** Whether this runtime reports effective permissions as authoritative state. */
val AgentRuntimeKind.reportsEffectiveThreadPermissions: Boolean
    get() = metadata?.capabilities?.reportsEffectiveThreadPermissions ?: true

/** Picker callers that only know `name` / `displayName` from a probe. */
fun isBetaAgentName(name: String, displayName: String): Boolean {
    val key = name.trim().lowercase()
    if (isStableAgentIdentity(key, displayName)) {
        return false
    }
    val cached = AgentRuntimeMetadataProvider.lookup?.invoke(key)
    return cached?.presentation?.isBeta ?: true
}

private fun isStableAgentIdentity(name: String, displayName: String): Boolean =
    name.trim().lowercase() == "codex" || displayName.trim().lowercase() == "codex"

private fun AgentRuntimeKind.titlecased(): String {
    if (isEmpty()) return "智能体"
    return substring(0, 1).uppercase() + substring(1)
}

/**
 * Renders an agent's icon from the local drawable catalog
 * (`R.drawable.agent_<id>`) when one is bundled, falling back to a
 * monogram letter chip. Use this everywhere — new alleycat-advertised
 * agents stay renderable without needing a baozi release first.
 */
@Composable
fun AgentIconView(
    kind: AgentRuntimeKind,
    sizeDp: Int = 24,
    modifier: Modifier = Modifier,
) {
    val context = LocalContext.current
    val resName = "agent_${kind.lowercase()}"
    val resId = context.resources.getIdentifier(resName, "drawable", context.packageName)
    if (resId != 0) {
        Image(
            painter = painterResource(id = resId),
            contentDescription = kind.runtimeLabel,
            modifier = modifier.size(sizeDp.dp),
        )
    } else {
        AgentMonogram(kind = kind, sizeDp = sizeDp, modifier = modifier)
    }
}

@Composable
fun AgentMonogram(
    kind: AgentRuntimeKind,
    sizeDp: Int = 24,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .size(sizeDp.dp)
            .clip(RoundedCornerShape((sizeDp * 0.2).dp))
            .background(Color.Black.copy(alpha = 0.82f))
            .border(
                width = 0.5.dp,
                color = BaoziTheme.textPrimary.copy(alpha = 0.25f),
                shape = RoundedCornerShape((sizeDp * 0.2).dp),
            ),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = kind.firstOrNull()?.uppercaseChar()?.toString() ?: "?",
            color = BaoziTheme.accent,
            fontSize = (sizeDp * 0.6).sp,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

@Composable
fun BetaBadge(modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .border(
                width = 0.5.dp,
                color = BaoziTheme.accent.copy(alpha = 0.6f),
                shape = RoundedCornerShape(3.dp),
            )
            .padding(horizontal = 5.dp, vertical = 1.dp),
    ) {
        Text(
            text = "BETA",
            color = BaoziTheme.accent,
            fontSize = 9.sp,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

File: apps/ios/Sources/Baozi/Views/AgentRuntimePresentation.swift
import SwiftUI

/// Bridge alias: Rust exposes agent identity as an opaque `String` (the
/// lowercase id alleycat advertises). The legacy `AgentRuntimeKind`
/// name is kept as a type alias so call sites compile; ALL agent
/// metadata — label, icon, BETA badge, sort order, capability flags —
/// comes from `AgentMetadataStore` keyed by id. There is no hardcoded
/// catalog of agent names in litter, so adding a new agent only
/// requires an entry in the alleycat manifest.
typealias AgentRuntimeKind = String

/// Lookup hook into the Rust-owned `AgentMetadataStore`. Wired up at
/// app launch in `BaoziApp` so any view can resolve an `AgentId` to
/// its metadata. Returns `nil` before the first probe response.
enum AgentRuntimeMetadataProvider {
    static var lookup: ((String) -> AppAgentMetadata?)?
    static var all: (() -> [AppAgentMetadata])?
}

extension AgentRuntimeKind {
    static let claude: AgentRuntimeKind = "claude"
    static let codex: AgentRuntimeKind = "codex"
    static let devin: AgentRuntimeKind = "devin"
    static let droid: AgentRuntimeKind = "droid"
    static let opencode: AgentRuntimeKind = "opencode"

    /// Presentation order surfaced by `AgentMetadataStore` (sorted by
    /// each agent's `presentation.sort_order` from the alleycat
    /// manifest). Empty when no probe has populated the cache yet —
    /// callers should treat that as "no agents available."
    static var presentationOrder: [AgentRuntimeKind] {
        AgentRuntimeMetadataProvider.all?().map(\.name) ?? []
    }

    var metadata: AppAgentMetadata? {
        AgentRuntimeMetadataProvider.lookup?(self)
    }

    /// Short label used in lists. Prefers metadata `display_name`;
    /// falls back to a titlecased id so the UI never shows a blank
    /// label during the brief window between server connect and probe
    /// completion.
    var displayLabel: String {
        if let meta = metadata, !meta.displayName.isEmpty {
            return meta.displayName
        }
        return titlecased
    }

    /// Header / title rendering. Prefers metadata `presentation.title`
    /// (e.g. "Factory Droid") over the short label.
    var titleDisplayLabel: String {
        if let title = metadata?.presentation?.title, !title.isEmpty {
            return title
        }
        return displayLabel
    }

    /// Sort index. Prefers metadata `sort_order`; otherwise drops to
    /// the end, tie-broken by name.
    var presentationSortIndex: Int {
        if let order = metadata?.presentation?.sortOrder {
            return Int(order)
        }
        return Int.max
    }

    /// BETA badge driven by `presentation.is_beta` from alleycat. Codex is
    /// always treated as stable, including cold-start SSH/alleycat paths where
    /// metadata may not be cached yet. Other unknown agents stay beta by
    /// default until metadata says otherwise.
    var isBeta: Bool {
        if Self.isStableAgentIdentity(self, displayName: "") {
            return false
        }
        return metadata?.presentation?.isBeta ?? true
    }

    /// Whether this runtime accepts client-side thread permission overrides.
    /// Older daemons did not advertise the capability, so default to the
    /// historical behaviour until a runtime explicitly opts out.
    var supportsThreadPermissionOverrides: Bool {
        metadata?.capabilities?.supportsThreadPermissionOverrides ?? true
    }

    /// Whether this runtime reports effective thread permissions that the UI
    /// can present as authoritative runtime state.
    var reportsEffectiveThreadPermissions: Bool {
        metadata?.capabilities?.reportsEffectiveThreadPermissions ?? true
    }

    /// Asset catalog name for this agent's bundled icon, by convention
    /// `agent_<id>`. Returns `nil` when no matching `UIImage(named:)`
    /// is bundled — callers fall back to a monogram chip via
    /// `AgentIconView`. Baozi ships icons for the agents it knows
    /// about (codex, claude, etc.) and renders a monogram for anything
    /// new that alleycat advertises.
    var bundledAssetName: String? {
        let candidate = "agent_\(self)"
        return UIImage(named: candidate) != nil ? candidate : nil
    }

    /// Picker / add-server callers check whether an agent should show
    /// a BETA badge before its metadata has been promoted into the
    /// store. With no enum to consult, defer entirely to the cached
    /// metadata; unknown agents are beta by default except Codex.
    static func isBetaAgentName(_ name: String, displayName: String) -> Bool {
        let key = name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if isStableAgentIdentity(key, displayName: displayName) {
            return false
        }
        return AgentRuntimeMetadataProvider.lookup?(key)?.presentation?.isBeta ?? true
    }

    private static func isStableAgentIdentity(_ name: String, displayName: String) -> Bool {
        name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "codex"
            || displayName.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "codex"
    }

    private var titlecased: String {
        guard !isEmpty else { return "Agent" }
        return prefix(1).uppercased() + dropFirst()
    }
}

/// Renders an agent's icon from the local asset catalog (`agent_<id>`)
/// when one is bundled, otherwise falls back to a monogram letter chip.
/// Use this everywhere instead of `Image(kind.assetName)` — it keeps
/// new alleycat-advertised agents renderable without shipping a litter
/// release first.
struct AgentIconView: View {
    let kind: AgentRuntimeKind
    var size: CGFloat = 24

    var body: some View {
        if let assetName = kind.bundledAssetName {
            Image(assetName)
                .resizable()
                .interpolation(.high)
                .frame(width: size, height: size)
        } else {
            AgentMonogramView(kind: kind, size: size)
        }
    }
}

/// Letter-based fallback when no icon is cached. Renders the first
/// character of the agent id in the accent color over a dark chip —
/// good enough for cold-start before the first probe completes.
struct AgentMonogramView: View {
    let kind: AgentRuntimeKind
    var size: CGFloat = 24

    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: size * 0.2)
                .fill(Color.black.opacity(0.82))
                .overlay(
                    RoundedRectangle(cornerRadius: size * 0.2)
                        .stroke(BaoziTheme.textPrimary.opacity(0.25), lineWidth: 0.5)
                )
            Text(monogramLetter)
                .font(.system(size: size * 0.6, weight: .semibold, design: .monospaced))
                .foregroundColor(BaoziTheme.accent)
        }
        .frame(width: size, height: size)
    }

    private var monogramLetter: String {
        kind.first.map { String($0).uppercased() } ?? "?"
    }
}

struct BetaBadge: View {
    var body: some View {
        Text("BETA")
            .baoziFont(.caption2)
            .foregroundColor(BaoziTheme.accent)
            .padding(.horizontal, 5)
            .padding(.vertical, 1)
            .overlay(
                RoundedRectangle(cornerRadius: 3)
                    .stroke(BaoziTheme.accent.opacity(0.6), lineWidth: 0.5)
            )
    }
}
